### PR TITLE
Add late gan training

### PR DESCRIPTION
**Pull request by Or-Tal**

### CODE REVIEW

File: denoiser/batch_solvers/adversarial_bs.py [hunk 1]
```diff
@@ -36,32 +38,58 @@ def run(self, data, cross_valid=False):
...
+                loss += F.smooth_l1_loss(clean, estimate)
+            else:
+                raise ValueError(f"Invalid loss {self.args.loss}")
+            # MultiResolution STFT loss
+            if self.args.stft_loss:
+                sc_loss, mag_loss = self.mrstftloss(estimate.squeeze(1), clean.squeeze(1))
+                loss += sc_loss + mag_loss
+
+            losses_dict = {self._losses_names[0]: loss.item(), self._losses_names[1]: 0}
+            losses = (loss, 0)
```
Comment (mosem) [inline denoiser/batch_solvers/adversarial_bs.py:73]: The whole calculation of losses for the first epochs seem like it's taken from the get_loss function of geneartor_bs, and since adversarial_bs inherits from generator_bs, can't you just call get_loss here instead?
The run function is pretty organized and short and I would like to keep it that way, so I prefer that if you can't call an existing function, create a dedicated function for this case and call it from here.

Also I would change the comment to "train all **the epochs before disc_first_epoch** simply...".

Since you are adding a different kind of loss, maybe we should rename get_total_generator_loss to get_adversarial_generator_loss, and get_discriminator_loss to get_adversarial_generator_loss? in order to better distinguish between the different phases of the training?

**Review by mosem — APPROVED** [review]
Looks great. I have one comment regarding the calculation of the loss in adversarial_bs.

File: denoiser/batch_solvers/adversarial_bs.py (hunk 1 above)
Comment (Or-Tal) [reply denoiser/batch_solvers/adversarial_bs.py:60]: ur right, replaced with super()._get_loss call.
changed the comment also.
I don't think the renaming is necessary, it is clear that we only train the generator for a given number of steps

Comment (mosem) [reply denoiser/batch_solvers/adversarial_bs.py:60]: Great. You can merge this branch.

**Review by mosem — APPROVED** [review]
